src/data_loader.cu

```diff
@@ -0,0 +1,239 @@
+#include <stdio.h>
+#include <stdint.h>
+#include <byteswap.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "../includes/data_loader.cuh"
+#include "../includes/common.cuh"
+
+
+void free_MNIST_images(MNISTImage *images, uint32_t num_samples) {
+    for (uint32_t i = 0; i < num_samples; ++i)
+        free(images[i].pixels);
+    free(images);
+}
+
+
+void free_MNIST_dataset(MNISTDataset *dataset) {
+    free_MNIST_images(dataset->images, dataset->num_samples);
+    free(dataset->labels);
+    free(dataset);
+}
+
+
+void free_images(Image *images, uint32_t num_samples) {
+    if (images == NULL)
+        return;
+    for (uint32_t i = 0; i < num_samples; ++i)
+        free(images[i].pixels);
+    free(images);
+}
+
+void free_dataset(ImageDataset *dataset) {
+    free_images(dataset->images, dataset->num_samples);
+    free(dataset->labels);
+    free(dataset->view_indices);
+    free(dataset);
+}
+
+
+uint8_t *_uint32_to_byte(uint32_t number) {
+    uint8_t *array = (uint8_t *)mallocCheck(4 * sizeof(uint8_t));
+    array[0] = (uint8_t)(number >> 24);
+    array[1] = (uint8_t)(number >> 16);
+    array[2] = (uint8_t)(number >>  8);
+    array[3] = (uint8_t)(number >>  0);
+    return array;
+}
+
+
+uint32_t _read_uint32_convert_endian(FILE *stream, const uint32_t invalid_return_value, const char *label) {
+    uint32_t number;
+    if (!fread(&number, sizeof(uint32_t), 1, stream)) {
+        printf("Error in reading the %s.\n", label);
+        return invalid_return_value;
+    }
+
+    return __bswap_32(number);
+}
+
+
+uint8_t _load_dimension_from_idx_file_header(FILE *stream) {
+    uint8_t invalid_return_value = 0;
+    uint32_t magic_number = _read_uint32_convert_endian(stream, invalid_return_value, "magic number");
+    if (magic_number == invalid_return_value)
+        return invalid_return_value;
+
+    // Separate the magic number into 4 one-byte components.
+    uint8_t *magic_byte_array = _uint32_to_byte(magic_number);
```

Comment:
Ah, this could be a cast vs an allocation, but let's say you do want a separate buffer to copy into, for this case, it could be better to do `uint8_t magic_byte_array[4]` and copy the values into it from `magic_number`.

A heap allocation tends to be seen as "expensive" perf wise if you don't need it. Since magic_byte_array is of a known size, and the lifetime is only within this function, we should take advantage of the stack instead of using the heap.